### PR TITLE
builddep: add feature to get builddeps from remote packages

### PR DESCRIPTION
**Pull request by ignatenkobrain**
Signed-off-by: Igor Gnatenko <i.gnatenko.brain@gmail.com>

### CODE REVIEW

**Comment by rholy** [comment]
See https://bugzilla.redhat.com/show_bug.cgi?id=1074585 for a discussion.

File: plugins/builddep.py [hunk 1]
```diff
@@ -62,9 +63,32 @@ def __exit__(self, exc_type, exc, exc_tb):
 class BuildDepCommand(dnf.cli.Command):
 
     aliases = ('builddep',)
-    msg = "Install build dependencies for .src.rpm, .nosrc.rpm or .spec file"
+    msg = "Install build dependencies for .src.rpm, .nosrc.rpm, .spec file or repo"
     summary = _(msg)
-    usage = _("[PACKAGE.src.rpm|PACKAGE.nosrc.rpm|PACKAGE.spec]")
+    usage = _("[PACKAGE.src.rpm|PACKAGE.nosrc.rpm|PACKAGE.spec|PACKAGE]")
```
Comment (jsilhan) [inline plugins/builddep.py]: we probably want to allow `<package-name-spec>` here [1] and do `get_best_selector`, etc.
[1] http://dnf.readthedocs.org/en/latest/command_ref.html#specifying-packages

File: plugins/builddep.py [hunk 2]
```diff
@@ -119,18 +143,27 @@ def _spec_deps(self, spec_fn):
...
-        demands.root_user = True
-        demands.sack_activation = True
-
-    @sink_rpm_logging()
-    def run(self, args):
-        rpm_ts = rpm.TransactionSet()
-        for fn in args:
-            if fn.endswith('.src.rpm') or fn.endswith('.nosrc.rpm'):
-                self._src_deps(rpm_ts, fn)
+    def _enable_source_repos(self):
```
Comment (jsilhan) [inline plugins/builddep.py]: move the method [1] from download plugin to lib adn import it, please
[1] https://github.com/rpm-software-management/dnf-plugins-core/blob/master/plugins/download.py#L161

**Comment by jsilhan** [comment]
and add `...(RhBug:1074585)` to commit message

File: plugins/builddep.py (hunk 1 above)
Comment (rholy) [inline plugins/builddep.py]: Yes, see the reference to Panu's comment in https://bugzilla.redhat.com/show_bug.cgi?id=1074585.

**Comment by rholy** [comment]
Because of the Panu's comment, I'm against this "feature".

File: plugins/builddep.py (hunk 1 above)
Comment (rholy) [inline plugins/builddep.py]: I mean no (I misread your comment).

**Comment by MichaelMraka** [comment]
Merged.

**Comment by rholy** [comment]
OK, so at least the Jan's request wasn't fulfilled and you've merged it, right? Hm...

**Comment by timlau** [comment]
This "feature" is broken by design as discussed in 
 https://bugzilla.redhat.com/show_bug.cgi?id=1074585

It should never have been merged

**Comment by MichaelMraka** [comment]
> OK, so at least the Jan's request wasn't fulfilled and you've merged it, right? Hm...

Which request do you have on mind?

**Comment by MichaelMraka** [comment]
> This "feature" is broken by design as discussed in https://bugzilla.redhat.com/show_bug.cgi?id=905697#c21

It's "broken" in the same way builddep from src.rpm is. There's no difference whether src.rpm is local or comes from repo. So either supporting both makes sense or we should better remove src.rpm option as well.

**Comment by rholy** [comment]
This is even more broken since, as Panu said, the downloaded SRPM is for an arbitrary arch which may have been build with different macros...

**Comment by MichaelMraka** [comment]
The same applies both for srpm downloaded by dnf from repo and srpm  downloaded by user and passed to dnf as local file.

**Comment by rholy** [comment]
> The same applies both for srpm downloaded by dnf from repo and 
This is not an excuse.
> srpm downloaded by user and passed to dnf as local file.
Not necessarily.

**Comment by MichaelMraka** [comment]
>> The same applies both for srpm downloaded by dnf from repo and

> This is not an excuse.

?!? I'm afraid I don't understand this comment.

>>  srpm downloaded by user and passed to dnf as local file.

> Not necessarily.

Could you elaborate more on this?

**Comment by timlau** [comment]
the mayor diffence by using a local .src.rpm and one fetched from the repo, is that the user selected the local one, but the user don't know what src.rpm he wil get in the second case, it depend a lot of the repo configuration.
It is right that 

dnf download foobar --source
yum builddep foobar*.src.rpm

will give the same result as your change, but there is lot of other ways to get .src.rpm files, like download from koji where you can get the specific .src.rpm build for the release you want.

**Comment by rholy** [comment]
>>> The same applies both for srpm downloaded by dnf from repo and

>> This is not an excuse.

>?!? I'm afraid I don't understand this comment.

The fact that our "download" plugin is broken does not excuse the attempt to break another plugin the same way.

>>>  srpm downloaded by user and passed to dnf as local file.

>> Not necessarily.

> Could you elaborate more on this?

I think Tim have described it pretty well.

**Comment by MichaelMraka** [comment]
> but there is lot of other ways to get .src.rpm files, like download from koji where you can get the specific .src.rpm build for the release you want.

Packages in source repos *are* built from koji so getting srpm from koji puts you into the exactly same trouble as a srpm from repo.
E.g. evolution-3.17.1-1.fc23.src.rpm in koji was built on arm - http://koji.fedoraproject.org/koji/taskinfo?taskID=9575140.

So chances srpm you download somwhere from the internet was built on a different arch is roughly equal to the same package from source repo. Therefore `dnf builddep package` makes no difference to `dnf builddep ./package.src.rpm`.

**Comment by MichaelMraka** [comment]
> The fact that our download plugin is broken does not excuse the attempt to break another plugin the same way.

What's broken in download plugin and how it applies to builddep plugin?

**Comment by rholy** [comment]
> So chances srpm you download somwhere from the internet was built on a different arch is roughly equal to the same package from source repo. Therefore dnf builddep package makes no difference to dnf builddep ./package.src.rpm.

Even if that's true, the SRPM still does not need to be download from an untrusted source. You can just build it locally. Or a colleague could build it locally. Or our nightly builds contains SRPMs build for correct architectures.

> What's broken in download plugin and how it applies to builddep plugin?

That it does not download the SRPM from which the given package was built. And that's the same problem that underlays this "discussion".

**Pull request closed without merging**